### PR TITLE
Restore basic functionality to the bytecode debugger

### PR DESCRIPTION
**Pull request by damiendoligez**
~~This is work in progress.~~ I still need to figure out what must be done to handle backtraces in the presence of fibers.
[edit: now it's ready]

### CODE REVIEW

File: runtime/debugger.c [hunk 1]
```diff
@@ -246,6 +246,16 @@ void caml_debugger_init(void)
...
+  /* We completely disable channel locking because:
+     1. The debugger is not compatible with threads, so we don't need locks.
+     2. This file doesn't use locks around channel operations, but
+        function [check_pending] in io.c assumes that the channel is locked
+        and tries to unlock it.
+  */
+  caml_channel_mutex_free = NULL;
+  caml_channel_mutex_lock = NULL;
+  caml_channel_mutex_unlock = NULL;
+  caml_channel_mutex_unlock_exn = NULL;
```
Comment (xavierleroy) [inline runtime/debugger.c]: I agree this should not make any difference for a single-domain single-threaded program.  Yet, I'd prefer if you would just take the locks on `dbg_in` and `dbg_out` just after they are created, and never release these locks.

(Note to self: investigate why `caml_ml_close_channel` does not lock the channel .)

**Comment by OlivierNicole** [comment]
I was also looking at this, but you are more advanced. If I can help with reviews or testing, let me know.

File: runtime/debugger.c (hunk 1 above)
Comment (damiendoligez) [reply runtime/debugger.c]: Ah yes, that's a much better solution, thanks.

File: runtime/debugger.c [hunk 2]
```diff
@@ -138,10 +138,16 @@ static void open_connection(void)
...
-     Lock and Unlock, but this is safe because the debugger only works
-     with single-threaded programs.  The program being debugged
-     will abort when it creates a thread. */
+  /* We need to lock the channels because the low-level I/O calls expect
+     to work on locked channels. We keep them locked because they are
+     private to this code and never used concurrently. */
+  Lock (dbg_in);
+  Lock (dbg_out);
+  /* We also need to remove them from the list of all open channels,
+     otherwise the flush-everything-at-exit code will try to relock them. */
```
Comment (xavierleroy) [inline runtime/debugger.c]: I'm surprised.  `caml_open_descriptor_{in,out}` does not set the `CHANNEL_FLAG_MANAGED_BY_GC` flag, so `caml_ml_out_channels_list` should not include these channels in its result list.  What's going on?

Comment (damiendoligez) [reply runtime/debugger.c:147]: I don't see any mention of `CHANNEL_FLAG_MANAGED_BY_GC` in `caml_{link,unlink}_channel`. The list of all open channels is used in `caml_ml_out_channels_list`, which is used to implement `Stdlib.flush_all`, which is itself used in `at_exit` to flush everything when the program exits.

Comment (damiendoligez) [reply runtime/debugger.c]: Oh I missed that. It's a strange interaction between memory management and `flush_all`.

Comment (xavierleroy) [reply runtime/debugger.c]: Well, there's the channels created from OCaml, which are subject to finalization and closed automatically when the program exits, and there's the channels created by the runtime system for its own uses (loading data from bytecode executables; communicating with the debugger), which are neither finalized nor auto-closed.  So it's not absurd to use a single flag `MANAGED_BY_GC` to govern both finalization and `flush_all`/`close_all`...

**Comment by xavierleroy** [comment]
This looks rather good to me.  Why don't you remove the Draft status so that it can get reviewed properly?

**Comment by xavierleroy** [comment]
Ping.  Why don't you remove the Draft status so that it can get reviewed properly?

**Comment by damiendoligez** [comment]
Stack backtrace is now working. The `next` instruction still behaves a little strange around fibers, but the debugger is already usable. I'll do another PR for `next` but I think it's less urgent.

**Review by xavierleroy — DISMISSED** [review]
Looks good to me.  Thanks a lot for this work.

I was able to run the demo program for effects from the OCaml 5 manual.  It's fun to step through all these continuations...

I added a commit that stops the debuggee cleanly if it spawns a domain.  This way we can still debug up to the point where a domain is created.  This is like #10594.

**Comment by fabbing** [comment]
A bit of code may have been lost during the PR and this should probably be added because the trap barrier doesn't seem to work anymore:
```diff
diff --git a/runtime/debugger.c b/runtime/debugger.c
index 187f681135..a220915784 100644
--- a/runtime/debugger.c
+++ b/runtime/debugger.c
@@ -560,6 +560,8 @@ void caml_debugger(enum event_kind event, value param)
       caml_flush(dbg_out);
       break;
     case REQ_SET_TRAP_BARRIER:
+      i = caml_getword(dbg_in);
+      CAMLassert(i == frame_block_number(frame_block));
       i = caml_getword(dbg_in);
       Caml_state->trap_barrier_off = -i;
       break;
```

I noticed the problem, thanks to @OlivierNicole, while trying to review this PR (I'm very new to bytecode and ocamldebug).

However, I don't think it's enough and `trap_barrier_off` should probably be a tuple `{ block, offset }`.
Even a tuple may still cause issue if we switch to a different fiber at the same depth (`block`), in this case raising an exception could trigger a trap barrier into the wrong fiber.

**Comment by damiendoligez** [comment]
@fabbing Thanks for finding this problem. This breaks the `finish` debugger command. I'm working on a solution.

**Comment by fabbing** [comment]
This is better than the "depth" in terms of number of fibers but will run into troubles if the fiber is reallocated since the `struct stack_info*` will change (see `caml_try_realloc_stack` from `runtime/fiber.c`).
As far as I know, there is currently no way to identify a fiber.
I was able to have a quick discussion on this topic with @kayceesrk and it seemed reasonable to add an unique identifier for each fiber. I'm working on a PR for this, I hope to have it ready soon.

**Comment by damiendoligez** [comment]
Adapted to use fiber IDs instead of addresses.
Another possibility would be to have `caml_try_realloc_stack` update `Caml_state->trap_barrier_block` as needed.

**Comment by fabbing** [comment]
I've resumed reviewing the PR and it looked good to me.
But I noticed that the `tool-debugger` tests from the testsuite where still skipped and tried to re-enable them again. All but one pass. The `dynlink` test fails with `Fatal error: Fatal error during unlock: Operation not permitted`.

It's not straightforward to debug such test as many fork are involved in the process.
I was able to pinpoint the issue at `channel_mutex_unlock_default` while attempting to create a `checkpoint` (`debugger.c:485`).
I don't understand why this only happens in this test and keep investigating!

**Comment by gasche** [comment]
With the OCaml 5 runtime you cannot safely call `fork` from C and then execute any OCaml code on the child that would release the runtime lock, you have to do a special dance first (and meditate on the sins of doing anything else than running an exec* function). The error I encountered in this situation is similar to yours, maybe look at #11111 to see if it is related; in your stead I would look at uses of `fork` in `runtime/debugger.c`.

File: runtime/debugger.c [hunk 3]
```diff
@@ -440,7 +481,13 @@ void caml_debugger(enum event_kind event, value param)
       break;
     case REQ_CHECKPOINT:
 #ifndef _WIN32
+      caml_release_domain_lock (); /* Don't fork while holding locks. */
+      Unlock (dbg_out);
+      Unlock (dbg_in);
       i = fork();
+      Lock (dbg_in);
+      Lock (dbg_out);
+      caml_acquire_domain_lock ();
```
Comment (gasche) [inline runtime/debugger.c:479]: Comparing this code with `caml_atfork_default` in domain.c, maybe you need to call `caml_reset_domain_lock` before re-acquiring the lock?

**Comment by fabbing** [comment]
Thanks a lot for the hint @gasche. I had a look in that direction but it turns out to be something different:

This test raises an exception (`Not_found` from `Sys.getenv`) which calls `Unlock_exn` / `channel_mutex_unlock_exn_default` and since `dbg_out` channel was locked last, causes it to be unlocked.
Later, when trying to create a `checkpoint`, an attempt is made to unlock `dbg_out` before forking which fails since it's no longer locked.

A not very graceful solution is to set `last_channel_locked` to `NULL` to avoid this:
```diff
--- a/runtime/caml/io.h
+++ b/runtime/caml/io.h
@@ -115,6 +115,8 @@ CAMLextern struct channel * caml_all_opened_channels;
 #define Flush_if_unbuffered(channel) \
   if (channel->flags & CHANNEL_FLAG_UNBUFFERED) caml_flush(channel)
 
+CAMLextern void caml_channel_forget_last(void);
+
 /* Conversion between file_offset and int64_t */
 
 #define Val_file_offset(fofs) caml_copy_int64(fofs)
diff --git a/runtime/debugger.c b/runtime/debugger.c
index 861c37e3bb..448098e2fb 100644
--- a/runtime/debugger.c
+++ b/runtime/debugger.c
@@ -143,6 +143,7 @@ static void open_connection(void)
      private to this code and never used concurrently. */
   Lock (dbg_in);
   Lock (dbg_out);
+  caml_channel_forget_last();
 
   if (!caml_debugger_in_use) caml_putword(dbg_out, -1); /* first connection */
 #ifdef _WIN32
@@ -479,6 +480,7 @@ void caml_debugger(enum event_kind event, value param)
       i = fork();
       Lock (dbg_in);
       Lock (dbg_out);
+      caml_channel_forget_last();
       caml_acquire_domain_lock ();
       if (i == 0) {
         close_connection();     /* Close parent connection. */
diff --git a/runtime/io.c b/runtime/io.c
index 78bfea6d8a..fae4d6239e 100644
--- a/runtime/io.c
+++ b/runtime/io.c
@@ -106,6 +106,11 @@ CAMLexport void (*caml_channel_mutex_unlock_exn) (void)
 /* List of opened channels */
 CAMLexport struct channel * caml_all_opened_channels = NULL;
 
+CAMLexport void caml_channel_forget_last(void)
+{
+  last_channel_locked = NULL;
+}
+
 /* Basic functions over type struct channel *.
    These functions can be called directly from C.
    No locking is performed. */
```
Thanks again to @OlivierNicole for his help in understanding channel behaviour.

Unfortunately, this `dynlink` test still fails because unable to place a breakpoint in `Plugin`:
```
Can't find any event there.
```
Investigation continues!

**Comment by damiendoligez** [comment]
This is strange. It says all tests have passed, but on my machine the dynlink test still fails.

**Comment by damiendoligez** [comment]
>Can't find any event there.

Fixed by 3b34535a65a675bfa0bbf6ce2b2f883973cff5b7, let's see if the testsuite passes now.

**Comment by fabbing** [comment]
Do we also want a "symmetrical" behavior with `resume` that would trigger a `trap` like `perform`?
```diff
--- a/runtime/interp.c
+++ b/runtime/interp.c
@@ -1284,6 +1284,7 @@ value caml_interprete(code_t prog, asize_t prog_size)
       goto do_resume;
 
 do_resume: {
+      check_trap_barrier_for_effect(domain_state);
       struct stack_info* stk = Ptr_val(accu);
       if (stk == NULL) {
          accu = Field(caml_global_data, CONTINUATION_ALREADY_TAKEN_EXN);
```
Either way it looks good to me!

**Comment by damiendoligez** [comment]
>Do we also want a "symmetrical" behavior with `resume` that would trigger a trap like `perform`?

I don't think so. `resume` is like a function call, it doesn't break the control flow of the current function. The trap barrier is here to make sure we intercept all the ways to get out of the current function, not just by returning but also with an exception or an effect. If we wanted to be exhaustive, we should add a trap barrier to the `exit` system call (and also `kill` with an unhandled signal on the current process) but that's probably not worth the trouble.

**Comment by damiendoligez** [comment]
@gasche, @xavierleroy can we get an official review ?

**Comment by xavierleroy** [comment]
> can we get an official review ?

Yes, but please leave me 1-2 weeks to work on it.  I'm thinking about a radical solution to the locking problems that have plagued this PR.  Stay tuned.

**Comment by xavierleroy** [comment]
Here is a preview of the "radical solution" I propose: https://github.com/damiendoligez/ocaml/pull/7

Ideally, I should make a separate PR with the basic changes, have it merged, and then you would rebase this PR on top of this work.  But first I'd like you (@damiendoligez , @fabbing , ...) to have a quick look at the approach and comment.

**Comment by xavierleroy** [comment]
I took the liberty to rebase on trunk and hand-merge https://github.com/damiendoligez/ocaml/pull/7.  I'm running a CI precheck test.

**Comment by damiendoligez** [comment]
precheck is at https://ci.inria.fr/ocaml/job/precheck/726/

**Review by xavierleroy — APPROVED** [review]
Test results are as positive as they can be, and the whole PR is in good shape as far as I can see, so let's merge!  Thanks @damiendoligez for all the work.

**Comment by xavierleroy** [comment]
Cherry-picked in 5.0: b4a61f5cba

**Pull request merged**